### DISCUSSION TITLE
Is it possible to map multiple inputs using MapInputName Attribute

### DISCUSSION THREAD

**Discussion by Thavarajan** [spatie/laravel-data#187] [Q&A]
```php
class ContractData extends Data  
{  
    public function __construct(
        public string $name,
        #[MapInputName('record_company')]
        public string $recordCompany,
    ) {
    }
}
```
for the above dataclass, i want to do the following
```php
$c = ContractData::from(['name'=>'test', 'record_company'=>'test']);
$c = ContractData::from(['name'=>'test', 'company'=>'test']);
```
the first one is fine, I want to transform the second one also, how can i convert the second one, with the same data class
  
if i want to map the 'record_company', 'company' to that recordCompany field, is it possible?

**Comment by Nir-An** [comment]
Hi  @Thavarajan 
It should be simple manipulation in ``fromArray`` method

something like :

```PHP
class ContractData extends Data  
{  
    public function __construct(
        public string $name,
        public string $recordCompany,
    ) {
    }

   public static function fromArray(array $contract)
  {
          return new self (
          name: $contract['name],
          recordCompany: $conrtact['company'] ? $conrtact['company'] : $conrtact['record_company'],
          // or just  $conrtact['record_company'] ?? $conrtact['company']
         }
  }
}
```
this is a simple example but the point is that you can solve it in a simple way
as I understand  MapInputName can work only for one input 

source : 
https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/creating-a-data-object#content-magical-creation

**Reply by Thavarajan** [reply]
Thanks for the response, wouldn't it duplicate everywhere,  I hope I need to right the same logic inside the fromrequest method also, and also the thing about validation rules

**Comment by rubenvanassche** [comment]
I don't think this is something we can implement easily, since we also need these names for validation. Which would we use then `company` or `record_company`? I think the solution @Nir-An provided is the best one is such cases.

**Reply by ragulka** [reply]
I don't understand why you would need to validate all the different input versions _before_ mapping them to the property? 

Given an example like this (assuming `MapInputName` can be made repeatable):

```php
class ContractData extends Data
{
    public function __construct(
        public string $name,
        #[MapInputName('record_company'), MapInputName('company')]
        public string $recordCompany,
    ) {
    }
}
```

And given a payload like `['name' => 'foo', 'company' => 'bar']`, I would assume the following:

1. Raw payload is passed to Data creation method
2. Raw payload is mapped according to mapping attributes, ie `company` is mapped to `recordCompany`
3. The payload now looks like `['name' => 'foo', 'recordCompany' => 'bar']`
4. The mapped payload is now validated
5. If validation passes, the Data object is created

**Reply by ragulka** [reply]
Going through the codebase, I think I may have misunderstood the purpose of input name mapping. I thought it would be useful for scenarios when the input may come from various different sources, and I can simply map them to my own internal property name. 

However, it seems its more geared towards mapping the input name for your own API endpoint, ie when you want to name your API fields different from your database fields. In this case, validating the mapped name does make sense, as the validation error would apply to the field name the API consumer would be using.

**Reply by rubenvanassche** [reply]
Actually maybe it is possible, but some changes to DataClass are required which is kinda breaking. But I think I would merge such a tested PR.

**Reply by eznix86** [reply]
@rubenvanassche Is this possible after all these years?

**Reply by rubenvanassche** [reply]
No PR, no feature 😅
